python312Packages.neurokit2: init at 0.2.10 and python312Packages.biosppy: init at 2.2.2

genga898 opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Closes #371375

Add [neurokit2](https://github.com/neuropsychology/NeuroKit) and [biosppy](https://github.com/scientisst/BioSPPy)

Tests for neurokit get stuck on the second test, have been trying to build it for couple of hours, hence the omission of the test checks, was not able to determine if all the tests pass or fail.

Biosppy is a dependency of the neurokit python lib.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bcdarwin commented on pkgs/development/python-modules/biosppy/default.nix:
```diff
@@ -0,0 +1,51 @@
+{
+  lib,
+  fetchFromGitHub,
+  python3Packages,
```
Please depend on individual package, not python3Packages.

bcdarwin commented on pkgs/development/python-modules/biosppy/default.nix:
```diff
@@ -0,0 +1,51 @@
+    homepage = "https://github.com/scientisst/BioSPPy";
+    license = lib.licenses.bsd3;
+    maintainers = with lib.maintainers; [ genga898 ];
+    mainProgram = "biosspy";
```
Is this a typo for `biosppy`?

bcdarwin commented on pkgs/development/python-modules/biosppy/default.nix:
```diff
@@ -0,0 +1,51 @@
+  meta = {
+    description = "Biosignal Processing in Python";
+    homepage = "https://github.com/scientisst/BioSPPy";
+    license = lib.licenses.bsd3;
```
Please add `changelog = https://github.com/scientisst/BioSPPy/releases/tag/v{version}`.

bcdarwin commented on pkgs/development/python-modules/biosppy/default.nix:
```diff
@@ -0,0 +1,51 @@
+    hash = "sha256-U0ZftAlRlazSO66raH74o/6eP1RpmuFoA6HJ+xmgKR8=";
+  };
+
+  build-system = with python3Packages; [
```
Please only add build-time deps to build-system and add the others to `dependencies`.

bcdarwin commented on pkgs/development/python-modules/biosppy/default.nix:
```diff
@@ -0,0 +1,51 @@
+    six
+    joblib
+    pywavelets
+    mock
```
Move test-time dependencies such as `mock` to `nativeCheckInputs`.

bcdarwin commented on pkgs/development/python-modules/biosppy/default.nix:
```diff
@@ -0,0 +1,51 @@
+    joblib
+    pywavelets
+    mock
+    sphinxHook
```
Doc dependencies such as `sphinx` should generally be omitted unless the package needs them at runtime.

bcdarwin commented on pkgs/development/python-modules/biosppy/default.nix:
```diff
@@ -0,0 +1,51 @@
+    recommonmark
+  ];
+
+  pythonRelaxDeps = true;
```
If there's only a few deps that need to be relaxed, better to list them.

bcdarwin commented on pkgs/development/python-modules/neurokit2/default.nix:
```diff
@@ -0,0 +1,56 @@
+    cvxopt
+    biosppy
+
+    # Test Dependencies
```
Same comments as for biosppy regarding `python3Packages`, test dependencies/`nativeCheckInputs`.

bcdarwin commented on pkgs/development/python-modules/neurokit2/default.nix:
```diff
@@ -0,0 +1,56 @@
+  meta = {
+    description = "Python Toolbox for Neurophysiological Signal Processing";
+    homepage = "https://github.com/neuropsychology/NeuroKit";
+    license = lib.licenses.mit;
```
Please add changelog similar to biosppy.

bcdarwin commented on pkgs/development/python-modules/neurokit2/default.nix:
```diff
@@ -0,0 +1,64 @@
+
+  build-system = [
+    setuptools
+    wheel
```
`wheel` is unlikely to be needed

bcdarwin commented on pkgs/development/python-modules/neurokit2/default.nix:
```diff
@@ -0,0 +1,64 @@
+    biosppy
+  ];
+
+  # Skipped tests cause they get stuck after the fifth one
```
In this case explicitly set `doCheck = false` followed by your comment

bcdarwin commented on pkgs/development/python-modules/biosppy/default.nix:
```diff
@@ -0,0 +1,58 @@
+  ];
+
+  nativeCheckInputs = [
+    mock
```
Without adding pytestCheckHook, unlikely the tests are actually run (also pytestCheckHook helpfully errors in the situation of no tests running)?

genga898 replied:
I added it, the tests run and then get stuck on the 5th test for almost an hour plus, I'm not really sure why.

bcdarwin commented on pkgs/development/python-modules/biosppy/default.nix:
```diff
@@ -0,0 +1,58 @@
+buildPythonPackage rec {
+  pname = "biosppy";
+  version = "2.2.2";
+
```
Please set `pyproject = true` (and `build-system = ...`) or similar

genga898 commented on pkgs/development/python-modules/neurokit2/default.nix:
```diff
@@ -0,0 +1,64 @@
+  build-system = [
+    setuptools
+    wheel
+    pytest_7
```
Pytest is required for it to build, it is required by `setuptools`

bcdarwin replied:
In any case no need to use an older pytest/pytestCheckHook.

bcdarwin replied:
Again, I think `pytest_7` should be replaced with pytest - I don't see any reason it's necessary to use this old version of pytest?

genga898 replied:
Ooh sorry for that I'll change the pytest version

bcdarwin reviewed (changes requested):
Typo in PR title: should be `neurokit2`.

After addressing these remaining issues, lease rewrite your commits so that one addresses biosppy and one addresses neurokit2.

I think it should be possible to enable the neurokit2 tests, potentially disabling one if it's hanging, although `mne` is broken right now; see #370107.

genga898 commented:
> Typo in PR title: should be `neurokit2`.
> 
> After addressing these remaining issues, lease rewrite your commits so that one addresses biosppy and one addresses neurokit2.
> 
> I think it should be possible to enable the neurokit2 tests, potentially disabling one if it's hanging, although `mne` is broken right now; see #370107.

I got the tests to work by disabling the failing ones and the one that was causing it to get stuck, the one's that pass do not meet the 50% mark, only at 22% meaning the build still fails

bcdarwin commented on pkgs/development/python-modules/neurokit2/default.nix:
```diff
@@ -0,0 +1,62 @@
+    biosppy
+  ];
+
+  # Passing tests do not meet coverage criteria
```
Note that Python packages shouldn't run `pytest-cov`. Not sure how it's getting into the closure here, but best to patch it out by adding `pytest-cov-stubs` to `nativeCheckInputs`. In that case, it could be possible to enable the tests.

bcdarwin commented on pkgs/development/python-modules/biosppy/default.nix:
```diff
@@ -0,0 +1,62 @@
+    pywavelets
+    mock
+  ];
+
```
I would still suggest explicitly adding `doCheck = false; # no tests` here for clarity.

bcdarwin commented on pkgs/development/python-modules/biosppy/default.nix:
```diff
@@ -0,0 +1,62 @@
+  ];
+
+  pythonImportsCheck = [
+    "biosppy"
```
note (optional): you can add additional modules/packages to `pythonImportsCheck` e.g. `biosppy.signals`, etc. to increase your confidence in the package being correctly packaged (especially in the absence of tests).

bcdarwin reviewed (changes requested):
Thanks, looks almost ready to go!

PerchunPak commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1677,6 +1677,8 @@ self: super: with self; {
 
   biopython = callPackage ../development/python-modules/biopython { };
 
+  biosppy = calPackage ../development/python-modules/biosppy { };
```
```suggestion
  biosppy = callPackage ../development/python-modules/biosppy { };
```

genga898 commented on pkgs/development/python-modules/neurokit2/default.nix:
```diff
@@ -0,0 +1,95 @@
+
+  build-system = [
+    setuptools
+    pytest
```
Yes

genga898 commented on pkgs/development/python-modules/neurokit2/default.nix:
```diff
@@ -0,0 +1,95 @@
+  ];
+
+  dependencies = [
+    scipy
+    scikit-learn
+    pandas
+    matplotlib
+    requests
+    cvxopt
+    biosppy
```
Not to be rude but wouldn't that just be a nitpick since it doesn't really affect anything.

genga898 replied:
I don't really see it anywhere in the python build docs

genga898 commented:
> Failing on darwin with:
> 
> ```
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
>   File "<string>", line 1, in <lambda>
>   File "/nix/store/8zc3wcplydp8gsxms24scpzdca438dk5-python3-3.12.8/lib/python3.12/importlib/__init__.py", line 90, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 999, in exec_module
>   File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
>   File "/nix/store/z21xzks0wrcxmz6xgx1pijzqiadi1k69-python3.12-biosppy-2.2.2/lib/python3.12/site-packages/biosppy/inter_plotting/__init__.py", line 18, in <module>
>     from . import ecg, acc
>   File "/nix/store/z21xzks0wrcxmz6xgx1pijzqiadi1k69-python3.12-biosppy-2.2.2/lib/python3.12/site-packages/biosppy/inter_plotting/ecg.py", line 20, in <module>
>     from tkinter import *
>   File "/nix/store/8zc3wcplydp8gsxms24scpzdca438dk5-python3-3.12.8/lib/python3.12/tkinter/__init__.py", line 38, in <module>
>     import _tkinter # If this fails your Python may not be configured for Tk
>     ^^^^^^^^^^^^^^^
> ModuleNotFoundError: No module named '_tkinter'
> ```

I could add it as an optional dependency on darwin builds but I don't have a machine to test for other failures

genga898 commented:
> > I could add it as an optional dependency on darwin builds but I don't have a machine to test for other failures
> 
> Give it a try, I can test it for you. If it still doesn't work, we can add:
> 
> ```nix
>   badPlatforms = [
>     lib.systems.inspect.patterns.isDarwin
>   ];
> ```
Just pushed an update, let me know if it works then I clean up my commit tree

genga898 commented on pkgs/development/python-modules/neurokit2/default.nix:
```diff
@@ -0,0 +1,95 @@
+    astropy
+    coverage
+    pytestCheckHook
+  ];
+
+  disabledTestPaths = [
```
Done

genga898 commented:
Ok let me fix my commit tree

genga898 committed:
python312Packages.biosppy: init at 2.2.2

genga898 committed:
python312Packages.neurokit2: init at 0.2.10
